core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -188,14 +188,13 @@ class DAGScheduler(
     eventProcessLoop.post(TaskSetFailed(taskSet, reason))
   }
 
-  private def getCacheLocs(rdd: RDD[_]): Array[Seq[TaskLocation]] = cacheLocs.synchronized {
+  private def getCacheLocs(rdd: RDD[_]): Seq[Seq[TaskLocation]] = cacheLocs.synchronized {
     // Note: this doesn't use `getOrElse()` because this method is called O(num tasks) times
     if (!cacheLocs.contains(rdd.id)) {
       val blockIds = rdd.partitions.indices.map(index => RDDBlockId(rdd.id, index)).toArray[BlockId]
-      val locs = BlockManager.blockIdsToBlockManagers(blockIds, env, blockManagerMaster)
-      cacheLocs(rdd.id) = blockIds.map { id =>
-        locs.getOrElse(id, Nil).map(bm => TaskLocation(bm.host, bm.executorId))
-      }
+      val locs = blockManagerMaster.getLocations(blockIds)
```

Comment:
ok thanks for checking. I think the style can be rewritten to something slightly more clear, e.g.
```scala
val locs: /*Add explicit type here */ = blockManagerMaster.getLocations(blockIds).map { blockLocs =>
  blockLocs.map(bm => TaskLocation(bm.host, bm.executorId))
}
cacheLocs(rdd.id) = locs
```